File: Source/Remote Control Server/TCP.vb
Imports System
Imports System.IO
Imports System.Net
Imports System.Net.Sockets
Imports System.Text
Imports System.Threading
Imports System.Windows.Threading

Public Class TCP

    Public Const portReceive As Integer = 1925
    Public Const portSend As Integer = 1927

    Public Const buffer As Integer = 1024
    Public Const receiveTimeout As Integer = 2000

    Public Const retries As Integer = 3
    Public Const retryTimeout As Integer = 1100
    Public Const sendTimeout As Integer = 1000

    Public tcpClient As TcpClient
    Public tcpListener As TcpListener

    Public isListening As Boolean = False
    Private listenTimer As DispatcherTimer

    Public Sub initialize()
        Logger.add("Initializing TCP")
        startListener()
    End Sub

    Public Function sendData(ByRef command As Command) As Boolean
        Dim received As Boolean = False
        Try
            If command.data.Length < 300 Then
                Logger.add("Sending command: " & command.dataAsString())
            Else
                Logger.add("Sending long command")
            End If


            Dim startTime As Long = My.Computer.Clock.TickCount

            tcpClient = New System.Net.Sockets.TcpClient()
            tcpClient.Client.SendTimeout = sendTimeout
            tcpClient.Connect(command.destination, portSend)
            Dim networkStream As NetworkStream = tcpClient.GetStream()

            networkStream.Write(command.data, 0, command.data.Length)
            tcpClient.Close()
            received = True

            Dim endTime As Long = My.Computer.Clock.TickCount
            Dim difTime As Long = endTime - startTime

            Logger.add("Success (" & difTime.ToString & "ms needed)")
        Catch ex As Exception
            Logger.add("Unable to send command to " & command.destination & ":" & portSend)
            Logger.add(ex.ToString)
        End Try
        Return received
    End Function

    Public Sub sendDataRetry(ByRef command As Command)
        For i As Integer = 0 To retries - 1
            If sendData(command) Then
                Exit For
            End If
        Next
    End Sub

    Public Sub sendDataUntilReceived(ByRef command As Command)
        Dim received As Boolean
        Do Until received
            received = sendData(command)
            Thread.Sleep(retryTimeout)
        Loop
    End Sub

    Public Sub startListener()
        Try
            If Not isListening Then
                isListening = True
                tcpListener = New TcpListener(IPAddress.Any, portReceive)
                tcpListener.Start()

                listenTimer = New DispatcherTimer
                AddHandler listenTimer.Tick, AddressOf listenTimerTick
                listenTimer.Interval = New TimeSpan(10000)
                listenTimer.Start()

                Logger.add("TCP listener started at port " & portReceive)
            Else
                Logger.add("TCP listener already running at port " & portReceive)
            End If
        Catch ex As Exception
            isListening = False
            Logger.add("Error while starting TCP listener:")
            Logger.add(ex.ToString)
        End Try
    End Sub

    Public Sub stopListener()
        isListening = False
    End Sub

    Private Sub listenTimerTick(ByVal sender As Object, ByVal e As EventArgs)
        listen()

        If Not isListening Then
            listenTimer.Stop()
        End If
    End Sub


    Public Sub listen()
        Try
            If tcpListener.Pending = True Then
                tcpClient = tcpListener.AcceptTcpClient
                tcpClient.ReceiveBufferSize = buffer

                Dim localEndPoint As Net.IPEndPoint = tcpClient.Client.LocalEndPoint
                Dim remoteEndPoint As Net.IPEndPoint = tcpClient.Client.RemoteEndPoint

                Dim Reader As New BinaryReader(tcpClient.GetStream)
                Try
                    Dim message_data As Byte()
                    message_data = Reader.ReadBytes(tcpClient.ReceiveBufferSize)

                    Dim command As New Command
                    command.source = remoteEndPoint.Address.ToString
                    command.destination = localEndPoint.Address.ToString
                    command.data = message_data
                    command.process()

                    'Logger.add("LocalEndPoint: " & command.destination)
                    'Logger.add("RemoteEndPoint: " & command.source)

                    Reader.Close()
                Catch ex As Exception
                    Logger.add("Error while parsing TCP data:")
                    Logger.add(ex.ToString)
                End Try
            End If
        Catch ex As Exception
            Logger.add("Error while receiving TCP data:")
            Logger.add(ex.ToString)
        End Try
    End Sub

End Class

File: Source/MacOS/RemoteControlServer/tcp.cpp
#include "logger.h"
#include "tcp.h"
#include "tcpthread.h"

#include <QByteArray>
#include <QTcpSocket>
#include <QThread>

#include <QDebug>
#include <QtConcurrent>

TCP* TCP::instance = NULL;

TCP* TCP::Instance()
{
    if (!instance)
    {
        instance = new TCP();
    }
    return instance;
}

TCP::TCP(QObject *parent) :
    QTcpServer(parent),
    portReceive(1925),
    portSend(1927),
    buffer(1024),
    receiveTimeout(2000),
    retries(3),
    retryTimeout(1100),
    sendTimeout(1000)
{
    isListening = false;

    Logger::Instance()->add("Initializing TCP");
    startListener();
}

void TCP::sendData(Command *command)
{
    QtConcurrent::run(this, &TCP::sendDataThread, command);
}

bool TCP::sendDataThread(Command *command) {
    bool received = false;
    QTcpSocket *socket = new QTcpSocket(0);
    socket->connectToHost(command->destination, portSend);

    if (socket->waitForConnected(sendTimeout))
    {
        socket->write(*command->data);
        received = true;
        socket->close();
    }
    else
    {
        Logger::Instance()->add("Unable to send command to " + command->destination + ":" + QString::number(portSend));
    }
    return received;
}

void TCP::sendDataRetry(Command *command)
{
    QtConcurrent::run(this, &TCP::sendDataRetryThread, command);
}

void TCP::sendDataRetryThread(Command *command)
{
    for (int i = 0; i < retries; ++i)
    {
        if (sendDataThread(command))
        {
            break;
        }
    }
}

void TCP::sendDataUntilReceived(Command *command)
{
    QtConcurrent::run(this, &TCP::sendDataUntilReceivedThread, command);
}

void TCP::sendDataUntilReceivedThread(Command *command)
{
    bool received = false;
    while (!received)
    {
        received = sendDataThread(command);
        QThread::sleep(retryTimeout);
    }
}

void TCP::startListener()
{
    if (!isListening)
    {
        isListening = true;
//        tcpServer = new QTcpServer();
//        bool success = tcpServer->listen(QHostAddress::Any, portReceive);
        bool success = this->listen(QHostAddress::Any, portReceive);
        if (success)
        {
//            connect(tcpServer, SIGNAL(newConnection()), this, SLOT(listenTimerTick()));
            Logger::Instance()->add("TCP listener started at port " + QString::number(portReceive));
        }
        else
        {
            isListening = false;
            Logger::Instance()->add("Error while starting TCP listener");
        }
    }
    else
    {
        Logger::Instance()->add("TCP listener already running at port " + QString::number(portReceive));
    }
}

void TCP::stopListener()
{
    this->close();
}

void TCP::listenTimerTick()
{
//    listen();
}

void TCP::incomingConnection(qintptr socketDescriptor)
{
    // We have a new connection
//    qDebug() << socketDescriptor << " Connecting...";

    TCPThread *thread = new TCPThread(socketDescriptor, this);

    // connect signal/slot
    // once a thread is not needed, it will be beleted later
    connect(thread, SIGNAL(finished()), thread, SLOT(deleteLater()));

    thread->start();
}

